capa/features/extractors/ts/engine.py

```diff
@@ -0,0 +1,214 @@
+
+    def get_new_object_names(self, node: Node) -> List[Tuple[Node, str]]:
+        return self.query.new_object_name.captures(node)
+
+    def get_assigned_property_names(self, node: Node) -> List[Tuple[Node, str]]:
+        return self.query.assigned_property_name.captures(node)
+
+    def get_function_definitions(self, node: Node = None) -> List[Tuple[Node, str]]:
+        return self.query.function_definition.captures(node if node is not None else self.tree.root_node)
+
+    def get_function_definition_name(self, node: Node) -> Node:
+        return node.child_by_field_name(self.query.function_definition_field_name)
+
+    def get_function_definition_names(self, node: Node) -> Iterator[Node]:
+        for fn_node, _ in self.get_function_definitions(node):
+            yield self.get_function_definition_name(fn_node)
+
+    def get_function_call_names(self, node: Node) -> List[Tuple[Node, str]]:
+        return self.query.function_call_name.captures(node)
+
+    def get_string_literals(self, node: Node) -> List[Tuple[Node, str]]:
+        return self.query.string_literal.captures(node)
+
+    def get_integer_literals(self, node: Node) -> List[Tuple[Node, str]]:
+        return self.query.integer_literal.captures(node)
+
+    def get_namespaces(self, node: Node = None) -> List[Tuple[Node, str]]:
+        return self.query.namespace.captures(node if node is not None else self.tree.root_node)
+
+    def get_global_statements(self) -> List[Tuple[Node, str]]:
+        return self.query.global_statement.captures(self.tree.root_node)
+
+
+class TreeSitterTemplateEngine(TreeSitterBaseEngine):
+    query: TemplateQueryBinding
+    language_toolkit: LanguageToolkit
+    embedded_language: str
+
+    def __init__(self, buf: bytes):
+        super().__init__(LANG_TEM, buf)
+        self.embedded_language = self.identify_language()
+        self.language_toolkit = LANGUAGE_TOOLKITS[self.embedded_language]
+        self.template_namespaces = set(name for _, name in self.get_template_namespaces())
+
+    def get_code_sections(self) -> List[Tuple[Node, str]]:
+        return self.query.code.captures(self.tree.root_node)
+
+    def get_parsed_code_sections(self) -> Iterator[TreeSitterExtractorEngine]:
+        for node, _ in self.get_code_sections():
+            # TODO: support JS
+            if self.embedded_language == LANG_CS:
+                yield TreeSitterExtractorEngine(
+                    self.embedded_language,
+                    self.get_byte_range(node),
+                    node.start_byte,
+                    self.template_namespaces,
+                )
+
+    def get_content_sections(self) -> List[Tuple[Node, str]]:
+        return self.query.content.captures(self.tree.root_node)
+
+    def identify_language(self) -> str:
+        for node, _ in self.get_code_sections():
+            if self.is_c_sharp(node):
+                return LANG_CS
+        return LANG_JS
```

Comment:
I believe there is no easy way to remedy this. From what I understand about templates in general is that the syntax is determined by the templating engine. In other words, there is easy way to detect from an unknown template which templating engine is being used (asp.net (and if so, what language), razor, ejs, erb, mako, jinja2, django, cheetah, go's html/template etc., not to mention each has their own syntax (some might use regular programming languages like C# to embed server logic, some might just contain very rudimentary placeholders/logic.

Here I am assuming that we only support EJS and C# in ASPX at the moment as embedded templates. This is because Tree-sitter embedded templates parser can only parse EJS and ERB (and we are not interested in embedded Ruby at the moment as far as I'm concerned). What's more, the default language for ASPX is VB, therefore if anyone wants to use C#, they need to include a \@Page directive with a Language attribute (see: https://docs.microsoft.com/en-us/previous-versions/aspnet/k33801s3(v=vs.100), https://docs.microsoft.com/en-us/previous-versions/dotnet/netframework-4.0/ydy4x04a(v=vs.100)?redirectedfrom=MSDN, https://docs.microsoft.com/en-us/previous-versions/aspnet/fbdt8kk7(v=vs.100)).

Reply:
Should we still assume that it's JS whenever it's not CS?
Could raise an Exception instead or are there other safe-guards in place before we get here?